yt/utilities/lib/cyoctree.pyx

```diff
@@ -1,679 +1,756 @@
-    np.int64_t leaf_id              # This is used in depositions (maybe)
-    unsigned char depth
+np.import_array()
+
+cdef struct Octree:
+    # Array of 3*num_nodes [x1, y1, z1, x2, y2, z2, ...]
+    np.float64_t * node_positions
+    # 1 or 0 of whether the oct has refined to make children
+    np.uint8_t * refined
+    # Each oct stores the depth in the tree: the root node is 0
+    np.uint8_t * depth
+    # pstart and pend tell us which particles in the pidx are stored in each oct
+    np.int64_t * pstart
+    np.int64_t * pend
+    # This tells us the index of each child
+    np.int64_t * children
+
+    # Here we store the coordinates and IDs of all the particles in the tree
+    np.float64_t * pposx
+    np.float64_t * pposy
+    np.float64_t * pposz
+    np.int64_t * pidx
+
+    # The max number of particles per leaf, if above, we refine
+    np.int64_t n_ref
+
+    # The number of particles in our tree, e.g the length of ppos and pidx
+    np.int64_t num_particles
+
+    # The total size of the octree (x, y, z)
+    np.float64_t * size
+
+    # The current number of nodes in the octree
+    np.int64_t num_nodes
+
+    # The maximum depth before we stop refining, and the maximum number of nodes
+    # we can fit in our array
+    np.uint8_t max_depth
+    np.int64_t max_num_nodes
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef int octree_build_node(Octree * tree, long int node_idx):
+    """
+    This is the main function in the building of the octree. This function takes
+    in the tree and an index of a oct to process. If the node has too many
+    particles (> n_ref) and the depth is less than the maximum tree depth then
+    we refine.
+
+    The `refine` creates 8 sub octs, within our oct. We indenfity the particles
+    in each of the subocts. We then recursively call this function on each of
+    the subocts.
+
+    Args:
+        tree: A pointer to the octree
+        node_idx: The index of the current node we are processing
+
+    Returns:
+        0
+    """
+    cdef np.int64_t splits[9]
+    cdef np.int64_t i, j, k, n, start, end
+    cdef np.float64_t lx, ly, lz, sz
+
+    if ((tree.pend[node_idx] - tree.pstart[node_idx] > tree.n_ref) and
+            (tree.depth[node_idx] < tree.max_depth)):
+        # If we are running out of space in our tree, then we *try* to
+        # relloacate a tree of double the size
+        if tree.num_nodes > tree.max_num_nodes - 16:
+            octree_reallocate(tree, tree.max_num_nodes * 2)
+
+        tree.refined[node_idx] = 1
+
+        # As we have decided to refine, we need to know which of the particles
+        # in this oct will go into each of the 8 child octs
+        split_helper(tree, node_idx, splits)
+
+        # Figure out the size of the current oct
+        sx = tree.size[0] / (2**tree.depth[node_idx])
+        sy = tree.size[1] / (2**tree.depth[node_idx])
+        sz = tree.size[2] / (2**tree.depth[node_idx])
```

Comment:
For performance reason, shouldn't this be stored in a temporary variable? Note that the following suggestion will also require defining `one_over_twotodepth`, and I am wondering whether we shouldn't use `1. / 2**tree.depth[node_idx]` to be on the safe side and prevent this from being interpreted as integer division?
```suggestion
        one_over_twotodepth = 1 / 2**tree.depth[node_idx]
        sx = tree.size[0] * one_over_twotodepth
        sy = tree.size[1] * one_over_twotodepth
        sz = tree.size[2] * one_over_twotodepth
```